File: 0572-Subtree of Another Tree/0572-Subtree of Another Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSubtree(TreeNode* s, TreeNode* t) {
        if (s == nullptr || t == nullptr) {
            return s == t;
        }
        
        return isSameTree(s, t) || isSubtree(s->left, t) || isSubtree(s->right, t);
    }
    
private:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (p == nullptr || q == nullptr) {
            return p == q;
        }
        
        return p->val == q->val && isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
};

File: 0501-0600/0572-Subtree of Another Tree/0572-Subtree of Another Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:
        def isSame(s, t):
            if s is None or t is None:
                return s == t
            return s.val == t.val and isSame(s.left, t.left) and isSame(s.right, t.right)
        
        if s is None or t is None:
            return s == t
                
        return isSame(s, t) or self.isSubtree(s.left, t) or self.isSubtree(s.right, t)
